### PR TITLE
Entered data missing when entering the wrong date for from, to in cart rule

### PR DESCRIPTION
**Pull request by realadityayadav**
### Description (*)
Changes in DataProvider and Save Controller for Persistance of Form Data In Cart Price Rule 
Updated Files - 
  -> app/code/Magento/SalesRule/Controller/Adminhtml/Promo/Quote/Save.php
  -> app/code/Magento/SalesRule/Model/Rule/DataProvider.php

### Fixed Issues (if relevant)

1. magento/magento2#20888 : Entered data missing when entering the wrong date for from, to in cart rule
2. ...

### Manual testing scenarios (*)

1 - Admin -> Marketing -> Cart Rule -> Add New
2 - Fill the all requird item
3 - Enter 'from'(Field) date grater than(>) 'to'(Field) date

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-engcom-team** [comment]
Hi @adityayadav-cedcoss. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me 2.3-develop instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

File: app/code/Magento/SalesRule/Controller/Adminhtml/Promo/Quote/Save.php [hunk 1]
```diff
@@ -39,6 +46,7 @@ public function __construct(
         $this->timezone =  $timezone ?? \Magento\Framework\App\ObjectManager::getInstance()->get(
             TimezoneInterface::class
         );
+        $this->dataPersistor = $dataPersistor;
```
Comment (sivaschenko) [inline app/code/Magento/SalesRule/Controller/Adminhtml/Promo/Quote/Save.php]: Please add ` $dataPersistor` as an argument to constructor

File: app/code/Magento/SalesRule/Controller/Adminhtml/Promo/Quote/Save.php [hunk 2]
```diff
@@ -20,13 +21,19 @@ class Save extends \Magento\SalesRule\Controller\Adminhtml\Promo\Quote implement
      * @var TimezoneInterface
      */
     private $timezone;
+    
+    /**
+     * @var DataPersistorInterface
+     */
+    protected $dataPersistor;
```
Comment (sivaschenko) [inline app/code/Magento/SalesRule/Controller/Adminhtml/Promo/Quote/Save.php]: please change the access level to private

File: app/code/Magento/SalesRule/Model/Rule/DataProvider.php [hunk 3]
```diff
@@ -35,6 +36,11 @@ class DataProvider extends \Magento\Ui\DataProvider\AbstractDataProvider
      * @var \Magento\SalesRule\Model\Rule\Metadata\ValueProvider
      */
     protected $metadataValueProvider;
+    
+    /**
+     * @var DataPersistorInterface
+     */
+    protected $dataPersistor;
```
Comment (sivaschenko) [inline app/code/Magento/SalesRule/Model/Rule/DataProvider.php]: please change access level to private

File: app/code/Magento/SalesRule/Model/Rule/DataProvider.php [hunk 4]
```diff
@@ -55,13 +62,15 @@ public function __construct(
         CollectionFactory $collectionFactory,
         \Magento\Framework\Registry $registry,
         \Magento\SalesRule\Model\Rule\Metadata\ValueProvider $metadataValueProvider,
+        DataPersistorInterface $dataPersistor,
```
Comment (sivaschenko) [inline app/code/Magento/SalesRule/Model/Rule/DataProvider.php]: Please add new dependencies in a backward compatible way, see "Adding a parameter to Constructor" section of https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/#php

**Review by sivaschenko — CHANGES_REQUESTED** [review]
Hi @adityayadav-cedcoss , thanks for the contribution! There are a couple of minor changes that should be done. Please take a look at my code review comments

File: app/code/Magento/SalesRule/Controller/Adminhtml/Promo/Quote/Save.php (hunk 2 above)
Comment (realadityayadav) [reply app/code/Magento/SalesRule/Controller/Adminhtml/Promo/Quote/Save.php]: It is changed.

File: app/code/Magento/SalesRule/Model/Rule/DataProvider.php (hunk 3 above)
Comment (realadityayadav) [reply app/code/Magento/SalesRule/Model/Rule/DataProvider.php]: It is changed.

File: app/code/Magento/SalesRule/Controller/Adminhtml/Promo/Quote/Save.php (hunk 1 above)
Comment (realadityayadav) [reply app/code/Magento/SalesRule/Controller/Adminhtml/Promo/Quote/Save.php]: ### It is changed.

**Comment by realadityayadav** [comment]
@sivaschenko Please take a look at the updates made.

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=20895) <br/>All committers have signed the CLA.

File: app/code/Magento/SalesRule/Model/Rule/DataProvider.php (hunk 4 above)
Comment (sivaschenko) [inline app/code/Magento/SalesRule/Model/Rule/DataProvider.php]: Please add this dependency in a backward compatible way, see "Adding a parameter to Constructor" section of https://devdocs.magento.com/guides/v2.3/contributor-guide/backward-compatible-development/#php

**Comment by sivaschenko** [comment]
Thanks for updates @realadityayadav ! Please add new constructor parameter to `app/code/Magento/SalesRule/Model/Rule/DataProvider.php` in backward compatible way and sign the [CLA](https://cla.dev.magento.com/magento/magento2?pullRequest=20895). Then I will be able to process this PR

File: app/code/Magento/SalesRule/Model/Rule/DataProvider.php (hunk 4 above)
Comment (realadityayadav) [reply app/code/Magento/SalesRule/Model/Rule/DataProvider.php]: Added the necessary changes @sivaschenko

**Review by sivaschenko — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sivaschenko, thank you for the review.
[ENGCOM-4215](https://jira.corp.magento.com/browse/ENGCOM-4215) has been created to process this Pull Request

**Pull request merged**